modules/images/webp-uploads/load.php

```diff
@@ -340,3 +340,130 @@ function webp_uploads_wp_get_missing_image_subsizes( $missing_sizes, $image_meta
 }
 
 add_filter( 'wp_get_missing_image_subsizes', 'webp_uploads_wp_get_missing_image_subsizes', 10, 3 );
+
+/**
+ * Filters `the_content` to update images so that they use the preferred MIME type where possible.
+ *
+ * By default, this is `image/webp`, if the current attachment contains the targeted MIME
+ * type. In the near future this will be filterable.
+ *
+ * Note that most of this function will not be needed for an eventual core implementation as it
+ * would rely on `wp_filter_content_tags()`.
+ *
+ * @since n.e.x.t
+ *
+ * @see wp_filter_content_tags()
+ *
+ * @param string $content The content of the current post.
+ * @return string The content with the updated references to the images.
+ */
+function webp_uploads_update_image_references( $content ) {
+	// This content does not have any tag on it, move forward.
+	if ( ! preg_match_all( '/<(img)\s[^>]+>/', $content, $img_tags, PREG_SET_ORDER ) ) {
+		return $content;
+	}
+
+	$images = array();
+	foreach ( $img_tags as list( $img ) ) {
+		// Find the ID of each image by the class.
+		if ( ! preg_match( '/wp-image-([\d]+)/i', $img, $class_name ) ) {
+			continue;
+		}
+
+		if ( empty( $class_name ) ) {
+			continue;
+		}
+
+		// Make sure we use the last item on the list of matches.
+		$attachment_id = (int) end( $class_name );
+
+		if ( ! $attachment_id ) {
+			continue;
+		}
+
+		// Create an array if attachment_id has not been added to the images array.
+		if ( empty( $images[ $attachment_id ] ) ) {
+			$images[ $attachment_id ] = array();
+		}
+
+		$images[ $attachment_id ][] = $img;
```

Comment:
I'm still not sure about the benefits of this structure. Why not `$images[ $img ] = $attachment_id;`?

Related to my other comments, the call to `wp_get_attachment_metadata()` doesn't really have a notable performance cost due to caching. Also, the way your code is structure here, it will take more cycles:
* With your current `images[ $attachment_id ][] = $img;`, if the same `img` tag is included 10 times in a post, `webp_uploads_img_tag_update_mime_type()` will be called 10 times, and the same `str_replace` will happen 10 times, which will not be necessary, since the first time will already replace all of these same occurrences.
* With `$images[ $img ] = $attachment_id;`, if the same `img` tag is included 10 times in a post, `webp_uploads_img_tag_update_mime_type()` will still only be called once and `str_replace` will also only run once.
* Both approaches will bring the same end result, but the latter does it more performantly.

Reply:
By using this structure we avoid the need to do what core does here:

```php
	// Reduce the array to unique attachment IDs.
	$attachment_ids = array_unique( array_filter( array_values( $images ) ) );
```

This structure would ensure only unique attachment IDs are stored, and avoid the need to use other functions.

One thing I'm having second thoughts is the fact that we need an array to store all the `img` tags as you describe if we have one should work for the majority of the cases as that would replace all instances on the content. However, I'm thinking in the use case where the same attachment ID uses 2 different markup elements like when you embed a particular size in the content and then embed another size the current implementation would be able to support that use case as would traverse all `img` tags for that particular attachment ID however the suggested approach won't be able to handle the following scenario:

> The content contains 2 images, at the top one image uses a small image size, at the bottom of the content the same image the same image uses only bigger sizes where the smaller size was not referenced.

In the scenario described above only the last element tag would be updated leaving the first image tag without modifications, unless the image uses any of the images from the same image at the bottom of the content.

Reply:
@mitogh Either of the implementations works. Something that is missing from your current implementation though is that every array of `$images[ $attachment_id ]` may include duplicate tags. In that case, it will run the logic below for the same tag multiple times which wouldn't do anything, it would just run logic without even making more replacements, because the first run per tag would already have made the replacement. So there would need to be some sort of `array_unique()` call for every `$images[ $attachment_id ]` to avoid it.

Using WP core's existing approach avoids that `array_unique()` call since `$images[ $image ]` automatically ensures there are no duplicates. But again, in that case we need `$attachment_ids = array_unique( array_filter( array_values( $images ) ) );` like you're saying.

IMO no approach here is really more suitable than the other. Therefore I generally would be okay going with your suggested approach, however since we're building towards WP core, I'd say let's use the same approach that WP core follows, to avoid having a discussion about that later, which would distract from what we want to achieve here.